src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -22,18 +22,57 @@ public InstalledAndTransitivePackageFeed(IEnumerable<PackageCollectionItem> inst
             _transitivePackages = transitivePackages ?? throw new ArgumentNullException(nameof(transitivePackages));
         }
 
+        internal static PackageCollectionItem SelectTransitiveLatestPackage(IGrouping<string, PackageCollectionItem> groupedPackagesById, IEnumerable<PackageCollectionItem> installedPkgs)
+        {
+            if (groupedPackagesById == null)
+            {
+                throw new ArgumentNullException(nameof(groupedPackagesById));
+            }
+
+            if (installedPkgs == null)
+            {
+                throw new ArgumentNullException(nameof(installedPkgs));
+            }
+
+            IEnumerable<PackageCollectionItem> matchedOrigins = groupedPackagesById
+                .Where(x => x.PackageReferences
+                    .Cast<ITransitivePackageReferenceContextInfo>()
+                    .Any(trPr => trPr.TransitiveOrigins.Any(trOrigin => installedPkgs.Contains(trOrigin.Identity))));
+
+            if (matchedOrigins.Any())
+            {
+                // Return the latest transitive package with installed versions
+                return matchedOrigins.OrderByDescending(x => x.Version).First();
+            }
+
+            // Otherwise, fallback on what's availble
+            return groupedPackagesById.OrderByDescending(x => x.Version).First();
```

Comment:
this assumes that groupedPackagesById is not empty right?
Should we be using `FirstOrDefault`?

Reply:
That or test the argument for empty right after the null check?

Reply:
If you do make this FirstOrDefault you'll need to handle the null return value in the caller.

Reply:
Added conditions to ensure groupedPackagesById is not empty.